### PR TITLE
Add Color(color,alpha) Color constructor

### LINKED ISSUE #23786: Allow for quick-creating Colors from other Colors
(sorry if the title isn't clear, but I wasn't sure how to name it)

The issue is:
Godot 3.1 (or 3.0?) introduced color constants. You can now use `Color.red` for red etc. However, I sometimes need a translucent red. So I still need to use `Color(1, 0, 0, 0.5)`. This is minor inconvenience of course, but since the `Color.red` was introduced to avoid `Color(1, 0, 0)` then why not go further?

There are two "fixes" I can think of:
`Color.red.with_alpha(0.5)` - it's actually longer, but I'd prefer that for whatever reason. Of course shorter version is possible (and preferred?), like `Color.red.w_a(0)`. (or `Color.red.translucent(0.5)`)

Or `Color.red.a = 0.5` could magically return red with 0.5 alpha`

This might of course apply to other Color properties to quickly create colors from other colors, but with minor changes.

Well, this >_>

### PR DESCRIPTION
**Pull request by ghost**
This implements copy constructors for Color with optional alpha, allowing constant expressions like `Color(Color.crimson)` and `Color(Color.crimson, 0.5)`. Closes #23786.

I just copy-pasted the relevant code from #21348, so all credits are due to @aaronfranke.

### CODE REVIEW

**Comment by aaronfranke** [comment]
Thanks for checking up on this! I actually have [this branch here](https://github.com/aaronfranke/godot/tree/copy-constructors) (and [commit](https://github.com/aaronfranke/godot/commit/5058080f5dc11eec33a91a0d8855718e1fee0636)), which includes both color alpha and copy constructors, but I haven't submitted it as a PR yet due to waiting on an answer to [this](https://github.com/godotengine/godot/issues/24001#issuecomment-513492209).

If you want, you can take that branch and push it into this PR. It would save me the work of rebasing that branch onto master after this is merged, and then your PR would also close #24001. But don't do this until @akien-mga and/or @vnen comments that they think it's a good idea.

**Comment by isaacremnant** [comment]
Oh boy I just realized it the number didn't refer to the number of arguments, not sure why I thought Color(4)+alpha(1) gave init6. I skipped init5 because it seemed like Color(color) wasn't too useful for a constructor, but I can't add it if you want. Otherwise I'll change init6 to init5.

By the way for init3 and init4, could [this](https://github.com/godotengine/godot/blob/master/core/variant.cpp#L1704) be the constructor you're talking about in the comment?

**Review by aaronfranke — APPROVED** [review]
A review from @vnen with regards to my above comment would be nice, but otherwise this PR should be good.

File: core/variant_call.cpp
Comment (akien-mga) [inline core/variant_call.cpp]: These are generated automatically for any legal variant conversion by https://github.com/godotengine/godot/blob/f46bf8e721d17f263437f4786454401b52721bc6/core/variant_call.cpp#L1384-L1399

If you want to change that, you'd have to remove the legal conversion between int/String and Color, and add a custom constructor instead.

Comment (aaronfranke) [reply core/variant_call.cpp]: The main reason would be to have more descriptive argument names. I'm not an expert on Variant so I don't know if there are any big downsides to doing this.

Comment (isaacremnant) [reply core/variant_call.cpp]: Sorry somehow didn't see the update two weeks ago. Super busy these days so I don't think I can spend the time to do this and test.

**Comment by Calinou** [comment]
Out of curiosity, could this be implemented using methods that return new Colors based on existing colors? We could have the following methods:

- `Color().with_red(float red)`
- `Color().with_green(float green)`
- `Color().with_blue(float blue)`
- `Color().with_alpha(float alpha)`

These methods could then be chained to modify multiple channels at once.

File: core/variant_call.cpp
Comment (aaronfranke) [inline core/variant_call.cpp]: This function (and line 1940) is not relevant anymore after #34948, since copy constructors are now built-in. This should be removed, then init6 can be renamed to init5.

**Review by aaronfranke — CHANGES_REQUESTED** [review]
Sorry for all the complications, but this should be the last time you have to update this PR :)

PS: About the comments about init3/init4, not sure what should be done about that, if we want to replace the Variant conversions or just leave a note of why init3/4 are skipped?

**Comment by isaacremnant** [comment]
Sorry I don't log in too often these days, too busy at work. Can someone pick this up?

**Comment by Mickeon** [comment]
Hasn't this already been salvaged by https://github.com/godotengine/godot/pull/36379?

**Comment by aaronfranke** [comment]
Good catch, thanks @Mickeon :)

**Comment by Mickeon** [comment]
Oooh so the tag *is* actually meant to be removed, I can keep that in mind.

**Comment by aaronfranke** [comment]
Yes, this way [these search results](https://github.com/godotengine/godot/pulls?q=is%3Aclosed+label%3Asalvageable) only include PRs that are still salvageable.

It's likely that there's a lot of cleaning up required on that list.

**Pull request closed without merging**
